basics_of_sst-core/tutorial/Example04.md

```diff
@@ -0,0 +1,488 @@
+
+            // Register the subcomponent API.
+            //
+            SST_ELI_REGISTER_SUBCOMPONENT_API(Example04::ExampleSubComponent)
+
+            // Register the subcomponent.
+            //
+            SST_ELI_REGISTER_SUBCOMPONENT(
+                ExampleSubComponent,                // INSERT_CLASS_NAME
+                "example",                          // Element library.  Name of the .so.
+                "ExampleSubComponent",              // "INSERT_CLASS_NAME"
+                SST_ELI_ELEMENT_VERSION(1, 0, 0),   // 
+                "Example subcomponent",             // Subcomponent description
+                "SST::SubComponent"                 // "INSERT_FULL_PARENT_CLASS_NAME" or 
+                                                    // "INSERT_COMPLETE_NAMESPACE::INSERT_PARENT_CLASS_NAME"
+            )
+
+        private:
+            int count_ = 0;
+    };
+}
+
+#endif
+```
+Like a component definition, the first few lines
+
+```
+#include <sst/core/subcomponent.h>
+#include <sst/core/output.h>
+```
+
+specify include files needed from the SST core libraries.  The subcomponent.h file defines the SST base subcomponent from which all SST subcomponents are derived.  The output.h file contains definitions used for logging.
+
+Next the namespace is defined
+
+```
+namespace Example04
+{
+```
+
+and the subcomponent defined.
+
+```
+    class ExampleSubComponent : public SST::SubComponent
+    {
+        public:
+            // Legacy constructors included for background compatibility.
+            // Needed for V9, can be removed for V10.
+            //
+            ExampleSubComponent(Component *owningComponent) :
+                SubComponent(owningComponent)
+            { }
+            ExampleSubComponent(Component* comp, Params& params) :
+                SubComponent(comp)
+            { }
+
+            // Updated constructors.  No longer pass in a parent.
+            // Instead, use a component id.
+            //
+            ExampleSubComponent(ComponentId_t id) :
+                SubComponent(id)
+            { }
+            ExampleSubComponent(ComponentId_t id, Params& params) :
+                SubComponent(id)
+            { }
+
+            virtual ~ExampleSubComponent() {}
+            
+            // Clock handler.  This is the method called from the clock event.
+            //
+            virtual void clock(Cycle_t) {}
+
+            // Subcomponent public method for returning current count.
+            //
+            int getCount() { return count_++; }
+```
+
+As is the case for components and events, all subcomponents are derived from SST::SubComponent or one of its children.
+
+The constructor and destructor are a standard format and required to interface with the SST core functions.  
+
+Subcomponents still have clock handlers but they are not automatically called by the simulation.  Instead they are called from the subcomponent parent as part of the parent's clock event handling.
```

Comment:
Not quite true. SubComponents can register a clock handler if they wish or they can implement a function that the Component calls during its own clock handler (or any other time). You may want to have a Component call a SubComponent clock handler if the order in which those handlers are called matters, but if it doesn't matter, then you can register an independent handler. It does not even have to be at the same clock frequency as the parent component. And no clock is required if the subcomponent doesn't need it.